### DISCUSSION TITLE
Help for configuring USB audio.

### DISCUSSION THREAD

**Discussion by Kydix** [mikebrady/shairport-sync#1623] [Q&A]
Hello, I try to configure shairport to use my Bose companion 5.

i have this 3 devices detected and my bose is correctly detected as "Bose USB Audio" in alsamixer,

```
hardware output devices:

      "hw:Audio"

      "hw:Headphones"

      "hw:vc4hdmi"
```

but I think my conf file is wrong

```
alsa =
{
        output_device = "hw:Audio"; // the name of the alsa output device. Use "shairport-sync -h" to discover the names of ALSA hardware devices. Use "alsamixer" or "aplay" to find out the names of devices, mixers, etc.

//      mixer_control_name = "Audio"; // the name of the mixer to use to adjust output volume. No default. If not specified, no mixer is used and volume in adjusted in software.

//      mixer_control_index = 0; // the index of the mixer to use to adjust output volume. Default is 0. The mixer is fully identified by the combination of the mixer_control_name and the mixer_control_index, e.g. "PCM",0 would be such a specification.

//      mixer_device = "Bose USB Audio"; // the mixer_device default is whatever the output_device is. Normally you wouldn't have to use this.


```

**Comment by neoncobbler** [comment]
Same issue on the same BOSE Companion 5 system. 
I've had some luck getting my Pi to output sound to the BOSE system by using the command `# raspi-config` then navigating to "System" and then "Audio" and selecting the "BOSE USB Audio" option. Then the command `speaker-test` and it works when playing the test tone. However, when I point the shairport .conf file at the hardware device I get nothing.

**Comment by mikebrady** [comment]
Hmm, it might be worth installing [sps-alsa-explore](https://github.com/mikebrady/sps-alsa-explore/blob/master/README.md) to see what's there. Is it possible that the Bose companion 5 only does 48000 frames per second and not 44100, which is what Shairport Sync needs?

**Reply by neoncobbler** [reply]
Hi Mike,

Well, that made progress. Now the companion 5 system works with Shairport when I manually run it from the command line. However, when I add it to the Raspbian auto exec with the `systemctl` command and reboot the pi Shairport will start and advertise itself to iTunes and I can connect to it but no sound. The `speakr-test` command still works and still plays the test sound to the BOSE system. I'm going to try some other USB sound devices and see if I have any better results.

**Reply by mikebrady** [reply]
Thanks for the update. Is it possible that you are running the full desktop version of Raspberry Pi OS? If so, it includes PulseAudio which is only available when a user logs in; hence the fact that Shairport Sync runs from the command line, but not as a system service, for which PulseAudio does not provide service.

Again, if so, you could consider making Shairport Sync start when the user logs in. This means that you don't have a fully autonomous system, since you have to log in first.

Alternatively, if you want a completely autonomous system, consider using Raspberry Pi OS Lite. You'll have no GUI, but also no PulseAudio, and Shairport Sync does run as an automatically-starting system service.

**Reply by neoncobbler** [reply]
Hi again Mike,

I got it working :)

I’ve been using Raspbian lite from the beginning and configuring it over SSH. Turns out the issue was the`systemctl` command.

For some reason the systemctl command wasn’t fully executing shairport. And would execute it far enough for shairport to advertise itself on my network but not enough to playback sound through any output device.

I got a new SD card and flashed it with a new image of raspbian lite. Once I installed shairport and updated the .config file to point it at my speakers it worked. It now runs completely headless/autonomously.

I believe that SD card was going bad or the initial image file was corrupted. 

Thanks for your support!

**Reply by rayoleary** [reply]
Exact same experience here - thanks so much for providing details above re: systemctl, it's doing that for me too.

**Comment by petcol** [comment]
hi guys,

Help please, I've been using analog so 
hardware output devices:

      "hw:Audio"

      "hw:Headphones"

for ages and that's been fine, but, the analog inplut on one unit has a balance and volume issue, so I'd really like to try sending digital via usb....
If I can be so bold how to change hardware output devices:

      "hw:Audio"

      "hw:Headphones"

      "hw:vc4hdmi"
NOT HDMI but USB ?

**Reply by mikebrady** [reply]
Thanks for the post. Nothing special has to be done to use USB audio — you just need to identify the device’s name and then tell Shairport Sync about it using a setting or command line argument. You can identify it by listing the output devices before and after inserting the USB device. Then you can use `alsamixer` to get the name of the mixer to control the volume.

**Reply by petcol** [reply]
mike is there a simple way to pick up which USB device it's using, after Reading the Manual the Pioneer XW-AM3 says to bring it out of standby with the USB plugged in, but that didn't work for the Pi, does for the iPod etc.

**Reply by mikebrady** [reply]
@petcol, reading the manual for the Pioneer XW-AM3, its USB connection seems to be a dedicated port for an old iPod/iPhone with the wide 30-pin connection; it is not for a USB soundcard, which is what Shairport Sync would use. Besides, the Pioneer XW-AM3 looks like a USB host, not a USB peripheral.

What you could do instead is, if the Pi has an audio output jack, route Shairport Sync audio to the Pi's audio output jack and use a headphone-jack-to-headphone-jack cable (it's called a "stereo mini-plug cable" in the manual) to plug the output from the audio jack into the auxiliary input ("AUX IN") of the XW-AM3.

**Reply by petcol** [reply]
thanks yes it has been working fine using the 3.5 mm stereo analogue input, well except that the plug or the socket on the pi is not 100% so the right channel not always engages which is why I'm trying to use pure digital.

**Reply by petcol** [reply]
I just tried a new build follows your build notes exactly :-)

I had the HDMI to USB cable plugged into the Speaker rebooted;

Then installed the sps-alsa-explore 
pi@raspberrypi4:~/sps-alsa-explore $ ./sps-alsa-explore
> Device Full Name:    "hw:Headphones"
  Short Name:          "hw:0"
  This device seems suitable for use with Shairport Sync.
  Possible mixers:     "PCM",0                 Range: 106.38 dB
  The following rate and format would be chosen by Shairport Sync in "auto" mode:
     Rate              Format
     44100             S16_LE

Still only detects the Headphones, so I guess that's what I'm stuck with ?

**Reply by mikebrady** [reply]
Perhaps you would consider using `raspi-config` to direct output to the HDMI. BTW I’m not sure if the HDMI is good…

**Reply by petcol** [reply]
yeah I did try that no change...  I guess there is something specific with the Pioneer speaker or some strange stuff generic to it.  I can just plug in a Apple iOS device to it's USB and music happens, I don't have easy access to a Andriod phone but I'll try that just to isolate if it works also.

It does work sort of fine on the 3.5 mm analogue just get the typical noises you get with analogue and dodgy 3.5 mm connectors.  Not such a problem at low volume but as it can not control the playback device you have to have it preset, so quiet or mute isn't such, just crackle n pops.

**Comment by glenvan** [comment]
I found this thread after fixing a similar issue with my Kanto TUK's USB interface. Just checking to make sure I'm using the optimal solution!

I'd just like to add one more thing for future readers: since `shairport-sync` exits if the expected USB hardware isn't powered on and present when you start streaming (rightly so), I added an accommodation to `systemd` to get the service back up after a quick face-palm.

```sh
sudo systemctl stop shairport-sync
sudo systemctl edit shairport-sync
```

Add these lines between the commented lines (as indicated in those same comments) to enhance the existing `systemd` service without modifying it explicitly:

```
[Service]
Restart=always
RestartSec=3
```

Save changes, and execute:

```sh
sudo systemctl daemon-reload
sudo systemctl start shairport-sync
```

Works great for me, because I'm not too worried about filling up my `syslog` with repeated attempts before I remember to power-on my speakers (YMMV).
